near-sdk/src/store/unordered_set/iter.rs

```diff
@@ -0,0 +1,412 @@
+use super::{CryptoHasher, UnorderedSet};
+use crate::store::free_list::FreeListIndex;
+use crate::store::{free_list, LookupMap};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::iter::{Chain, FusedIterator};
+use std::marker::PhantomData;
+
+impl<'a, T, H> IntoIterator for &'a UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize + Clone,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    type Item = &'a T;
+    type IntoIter = Iter<'a, T, H>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// An iterator over elements of a [`UnorderedSet`].
+///
+/// This `struct` is created by the [`iter`] method on [`UnorderedSet`].
+/// See its documentation for more.
+///
+/// [`iter`]: UnorderedSet::iter
+pub struct Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    elements: free_list::Iter<'a, T>,
+
+    el: PhantomData<H>,
+}
+
+impl<'a, T, H> Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    pub(super) fn new(set: &'a UnorderedSet<T, H>) -> Self {
+        Self { elements: set.elements.iter(), el: Default::default() }
+    }
+}
+
+impl<'a, T, H> Iterator for Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        <Self as Iterator>::nth(self, 0)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.elements.size_hint()
+    }
+
+    fn count(self) -> usize {
+        self.elements.count()
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.elements.nth(n)
+    }
+}
+
+impl<'a, T, H> ExactSizeIterator for Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+}
+impl<'a, T, H> FusedIterator for Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+}
+
+impl<'a, T, H> DoubleEndedIterator for Iter<'a, T, H>
+where
+    T: BorshSerialize + Ord + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        <Self as DoubleEndedIterator>::nth_back(self, 0)
+    }
+
+    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
+        self.elements.nth_back(n)
+    }
+}
```

Comment:
I don't think a double-ended iterator impl is needed for this and doesn't make sense in this context because the set has no order. I know I added it for `UnorderedMap`, but I'm second-guessing myself here.

@matklad your opinion on this would be valuable, do you think `DoubleEndedIterator` impls should be removed from unordered collections? impls don't exist for `std` equivalents

Reply:
Hmm, that is a good point. It probably does not make sense to treat an unordered set as something that has a "front" and a "back".